enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -119,7 +121,12 @@ type HitTrackerFactory struct {
 }
 
 func (h *HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	// For Action Cache hit-tracking, explicitly ignore everything. These cache
+	// If not hitting the remote cache, use a hit-tracker that sends information
+	// about local cache hits to the RPC service at the configured backend.
+	if proxy_util.SkipRemote(ctx) {
```

Comment:
This check shouldn't be necessary, the calling code should not get into the action_cache_server if SkipRemote is true, so I'd just remove it for simplicity.

Reply:
Just to clarify - you mean we will never return the NoOpHitTracker when SkipRemote=false right? 

We need a HitTrackerClient here when SkipRemote=true, because we won't have a remote cache doing hit tracking

Reply:
Oh sorry, thinking about this a little bit more carefully, I think what you actually want is to
1. Add support for tracking misses in the hit-tracker RPC
2. Add a bit in the `HitTrackerClient` that causes it to track those misses or not, and set that to track-misses when `SkipRemote` is true, and to not-track-misses when `SkipRemote` is false. The idea here is we want to track misses for requests that are only served locally.

Reply:
My understanding was that we don't need `TrackMiss` for workflows, because the misses data is only used in the cache scorecard, which does not apply for workflows.

The way I attempted to implement this is that we will now hit `TrackUpload` and `TrackDownload` when`SkipRemote=true`, which should cover all the data we need tracked. Am I missing something in my understanding?

Reply:
even if you don't need it right now, it's unexpected to get a tracker that can't track misses.  this should at least be a TODO and a bug if we don't address it right now.

i thiiink i agree with @iain-macdonald that we should always return a `HitTrackerClient` here? what's the scenario where we call `NewACHitTracker` but don't want a Real one?

Reply:
Added a TODO. Slight preference to not implement now, because it's unneeded and there have been some performance issues with proxy hit tracking in the past

You're right that we should never hit this code, unless we want a real `HitTrackerClient`. Modified the code to emit an unexpected event alert if `SkipRemote=false`

Reply:
Sorry if I'm still missing something here, but I _thiiiiink_ this can be simplified to:
1. Always return a real hit tracker client in `NewACHitTracker` and `NewCASHitTracker`
2. Remove the unexpected event in `SetExecutedActionMetadata` with a comment explaining why it's safe to do nothing there.

That'd make this code a bit simpler IMO. Lemme know if I have something off in my understanding though?

Reply:
I think it's just a matter of how defensive we want to be. We should never hit the cases where those unexpected events are emitted. I added them so we'll be alerted if something unexpected occurs with hit tracking, or if someone refactors this code, it will force them to be more deliberate